keps/sig-scheduling/3521-pod-scheduling-readiness/README.md

```diff
@@ -827,6 +827,18 @@ Longer term, we may want to require automated upgrade/rollback tests, but we
 are missing a bunch of machinery and tooling and can't do that now.
 -->
 
+- Start a local Kubernetes 1.26 cluster (`PodSchedulingReadiness` defaulted to false)
+- Create a Pod `pause` with one scheduling gate
+- The Pod's scheduling gate gets dropped as expected due to disabled feature gate
+- Delete the Pod
+- Re-start API Server and scheduler with version 1.27, and specify `PodSchedulingReadiness=true`
+- Create the same Pod `pause` with one scheduling gate
+- The Pod stays in `SchedulingGated` state, and its `.spec.schedulingGate` is persisted
+- Re-start API Server and scheduler with version 1.26
+- The Pod `pause` enters `Pending` state, with old `.spec.schedulingGate` reserved
+- Re-start API Server and scheduler with version 1.27, and specify `PodSchedulingReadiness=true`
+- The Pod stays in `Pending` state, with old `.spec.schedulingGate` reserved
```

Comment:
Although it's not returning to its original state, it's somehow inevitable. Let me know if you're concerned with this.

Reply:
Hmm - that's not expected to me.
IIUC, the state "SchedulingGated" is marked by scheduler when it observes the pod has a scheduling gate and puts it into a separate internal structure. Is that right?
If so, why this isn't happening restart with FG reenabled?

Reply:
Let me explain a bit, it's a bit complicated... The literal `Pending` vs. `SchedulingGated` is from the "STATUS" column of kubectl, which evaluate the pod's status with type `PodScheduled` - check whether its reason is "SchedulingGated", or "Unschedulable", etc.

### Step 1: upgrade

The pod gets gated. It's a happy path. Everything works fine, the pod's status is applied with {PodScheduled: SchedulingGated}

### Step 2: upgrade -> downgrade

No gates out there, so scheduler tries to schedule and bind it. But as mentioned above, it ends up with binding failure, so its status is applied with:

```yaml
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: null
    message: 'running Bind plugin "DefaultBinder": Operation cannot be fulfilled on
      pods/binding "pause": pod pause has non-empty .spec.schedulingGates'
    reason: SchedulerError
    status: "False"
    type: PodScheduled
  phase: Pending
  qosClass: BestEffort
```

### Step 3: upgrade -> downgrade -> upgrade

The pod is still re-gated as expected. It's just its **condition** is not updated back to {PodScheduled: SchedulingGated}. So the "STATUS" column of kubectl still shows as `Pending`.

---

To mark it back as `SchedulingGated`, I think I need to relax the check here:

https://github.com/kubernetes/kubernetes/blob/d2f40481d115155f3ce7abc0b7b2ff2cf8a0bb1e/pkg/registry/core/pod/strategy.go#L661-L666

A bit late for me now, will take a close look tomorrow.

Reply:
Thanks for explanation.

Yeah - I think that pod strategy needs to be slightly updated to make it more intuitive.
But we have time for it until code freeze.
[BTW - I'm OOO next week, so won't be responsive.]

Reply:
> To mark it back as SchedulingGated, I think I need to relax the check here:

Neither storage.go#PrepareForCreate() nor PrepareForUpdate() is triggered for a persisted pod, in the downgrade->upgrade path. I'm gonna ask for some suggestions in api-machinery.

Reply:
Confirmed with Jordan on [slack](https://kubernetes.slack.com/archives/C0EG7JC6T/p1676058998907059?thread_ts=1676058472.750079&cid=C0EG7JC6T), currently there is no way to hook the check in the downgrade->upgrade path:

> PrepareForCreate/PrepareForUpdate only trigger when a write is done, and upgrades don't automatically trigger writes

Reply:
Catching up with that (sorry for delay)

So I'm fine with that, but I guess I would expect that scheduler with overwrite it soon after upgrade, no?

Reply:
For now, there is no machinery to detect whether it's an upgrade path (although Daniel mentioned he has a [draft](https://kubernetes.slack.com/archives/C0EG7JC6T/p1676065545873709?thread_ts=1676058472.750079&cid=C0EG7JC6T)). With that said, for a persisted pod with a {PodScheduled, SchedulerError} condition (possible in a upgrade/downgrade path), although it's literally still gated in scheduler's perspective, but in UX's perspective, it still shows as "Pending" (instead of "SchedulingGated" as which needs to trigger `storage.go#PrepareForCreate()` while cannot be hit in this scenario)

Reply:
I understand that.

My point was rather: "isn't scheduler periodically processing those and can potentially update that on one of those syncs after upgrade?"

Reply:
Regarding "periodical processing", yes, it does [exist](https://github.com/kubernetes/kubernetes/blob/c3e7eca7fd38454200819b60e58144d5727f1bbc/pkg/scheduler/internal/queue/scheduling_queue.go#L335). However, the design of PodSchedulingReadiness is to leave the condition update to [API server](https://github.com/kubernetes/kubernetes/blob/c3e7eca7fd38454200819b60e58144d5727f1bbc/pkg/registry/core/pod/strategy.go#L676-L681) instead of scheduler.

BTW: scheduler may seek to remove the so-called "periodical processing" entirely for unschedulable pods in the future.

Reply:
Late to the party, but it sounds different from what other controllers do to leave condition updates to apiserver.
Although, IIUC, the binding endpoint also updates conditions in behalf of the user.

If a pod has 2 scheduling gates and there is an update that removes 1, would apiserver put the condition back to gated?

Maybe we could also make the scheduler send a condition update if it's gated. It's similar to what we do when we update condition when a scheduling attempt fails to accommodate the pod.

Reply:
> Although, IIUC, the binding endpoint also updates conditions in behalf of the user.

That's true. For scheduling gates, one-directional mutation (deletion after creation) is allowed, so it was designed to impose the condition once, at its creation time - rather than being updated upon each PreEnqueue calling.

> If a pod has 2 scheduling gates and there is an update that removes 1, would apiserver put the condition back to gated?

- If the pod carries the {PodScheduled, SchedulerError} condition (i.e., in upgrade/downgrade path), apiserver won't.
- If it's a regular workflow, the pod's creation would enable it carrying the {PodScheduled, SchedulingGated} condition, via API server, and thus remove partial gate will keep the condition as-is.

All-in-all, the SchedulingGated condition is imposed at the pod's creation time.

> Maybe we could also make the scheduler send a condition update if it's gated.

To me, it's a tradeoff question: does it worth the extra complexity to adapt to the upgrade/downgrade path? I'm not sure about that.

Reply:
sounds good. updated the wordings.